### DISCUSSION TITLE
Lit Elements on Wordpress

### DISCUSSION THREAD

**Discussion by JacobWeinbren** [lit/lit#2139] [Q&A] [answered]
@hasanirogers [proposed to use lit element to build SPA Wordpress themes](http://blog.hasanirogers.me/2020/09/web-app-with-wordpress-and-litelement.html). It was a really good idea in my opinion so I wanted to try it too. This prompts my question - how do you use Lit Element and Polyfills using WordPress? The difficulty is that the [polyfills suggested](https://lit.dev/docs/tools/production/) by Lit Element transpile javascript onto HTML - not javascript onto javascript. The suggested route in the blog is to use a ES5 transpiler. I think this makes sense, however you are also requiring modern browsers to run ES5 code - further, modern browsers are loading in polyfills, and legacy browsers are loading in things like prefixes. Typically you do targeting to solve this problem.

So my thinking is the answer would go something like.

1. Create a modern bundle, including just your bundle of code, including assets etc. you would typically rollup, along with with postcss

2. Create a legacy bundle with
- The Babel polyfills bundle.
- Lit Element polyfill-support.js polyfill loader
- The SystemJS loader.
- The bundle of code

3. Run a script running client-side detection like '!!Array.prototype.flat' which would decide which of these bundles to load.

But there are some shortcomings to my answer because essentially you would be dynamically loading javascript (not advised) after running the client test, which would come after preload and this would defeat the quick load times of LitElement. This would probably even be slower than just loading in the babel polyfills initially.

What do you think would be the best route to load polyfills in this situation. Further, if the Lit Element devs are reading this [your old advice](https://lit-element.polymer-project.org/guide/build) on this topic was really helpful - what things do you think have changed since these docs?

Best,
Jacob

**Comment by thescientist13** [comment, marked as answer]
> the difficulty is that the polyfills suggested by Lit Element transpile javascript onto HTML - not javascript onto javascript.

Not quite sure I understand what you mean here, do you mind elaborating a bit or pointing out where in the doc that information is presented?

----

If I am understanding your post correct though, I _think_ what you're looking for actually spans a couple distinct, but related topics / steps.

_Polyfills_ should typically come first and should be unbundled (minified is fine) and follow the pattern of feature detection / progressive enhancement.  If the browser supports something like `window.customElements` then don't do anything, otherwise, load up that functionality. [_webcomponents-loader.js_](https://github.com/webcomponents/polyfills#getting-started) works by only loading the feature detection logic in a very slim file size, and _if_  it detects the browser is missing something, it will delegate and load in the appropriate WC polyills as needed.   This all needs to be done synchronously before any dependent code can come into play.

For bundling, a common technique to apply here is called [differential loading](https://web.dev/codelab-serve-modern-code/), which as you point is effectively creating two versions of your app code, one for modern browsers and one for legacy browsers.  Typically the cutoff here is done along the lines of the introduction of ESM to HTML.  Basically, if the browser understands `type="module"`, it will load that and ignore `nomodule`, otherwise the legacy browser will "ignore" the first since it doesn't know how to handle `type="module"`, and load the second instead and just ignore the `nomodule` attribute.
```html
<script type="module" src="main.js"></script>
<script nomodule src="main.bundle.js" defer></script>
```

----

So all together, your _index.html_ might look something like this
```html
<html>

  <head>
    <script src="webcomponents-loader.js"></script>    
    <script type="module" src="main.js"></script>
    <script nomodule src="main.es5.js" defer></script>
  </head>
  
  <body>
     <h1>Welcome!</h1>
     <my-component></my-component>
  </body>

</html>
```

> _If you don't care about IE11 which I think is more and more of reality these days for greenfield projects I think there is less of a need for differential loading.  If you want to be safe, I think a good approach today would be author in ES6 and target your browsers / users using a tool like [**browserslist**](https://github.com/browserslist/browserslist) and see if you can get away with only having to deliver one bundle.  (IMO)_

Hope that helps!  ✌️

**Comment by JacobWeinbren** [comment]
Thank you this has been a very helpful response - in the end, I did as the comment at the bottom suggests.
